frontends/PyCDE/src/pycde/types.py

```diff
@@ -42,17 +44,35 @@ def maybe_create_alias(self, inner_type, name):
       if name in self.registered_aliases:
         if alias != self.registered_aliases[name]:
           raise RuntimeError(
-              f"Re-defining type alias for {name}! "\
-              f"Given: {inner_type}, "\
-              f"existing: {self.registered_aliases[name].inner_type}"
-          )
+              f"Re-defining type alias for {name}! "
+              f"Given: {inner_type}, "
+              f"existing: {self.registered_aliases[name].inner_type}")
         return self.registered_aliases[name]
 
       self.registered_aliases[name] = alias
       return alias
 
     return inner_type
 
+  def declare_types(self, mod):
```

Comment:
Regarding the comment from the previous PR:

> You just have to index the cache by module as well as name.

I don't see a way to index the cache by module, for the same reasons we were discussing... the module might not exist by the time a type is created with a name. Rather, this method is parameterized by module.

In addition, there are new instance variables to track which scopes have been created, and which declarations have been emitted. _These_ are indexed by module, so we have can the same set of aliases emitted into different modules, while still de-duping within a given module (to make `declare_types` idempotent).

This won't really support having totally separate sets of aliases for different modules, but I'm honestly not sure if that's a problem we need to solve yet. I'm imagining this functionality being used within the confines of a PyCDE System, and at the moment I don't know of any use-cases with multiple top-level PyCDE Systems (that want totally separate type declarations). Is this sufficient for now? If not, I think the right approach might be to have the `_Types` instance created per-System, rather than once when `pycde.types` is imported.

Reply:
> the module might not exist by the time a type is created with a name. Rather, this method is parameterized by module.

Good point. As for supporting different type alias per mlir module -- I don't see how or why we'd want to support that. The pycde modules are declared at the before mlir modules are even created so I think users would accept type alias being the same in all mlir modules. 

What I do worry about is polluting this global cache. Just importing a user module shouldn't cause type aliases to register themselves globally and thus preclude other type aliases with the same name -- that's not good for composability. This is a bridge we don't have to cross yet, so I don't think we should worry about it yet.

Reply:
Fair enough. The more I think about it, when we do cross that bridge, the solution will probably have to do with dialing back the idea of one global _Types instance, and giving finer grained control over how and when instances get created.